### DISCUSSION TITLE
Clarification on EIP 3860

### DISCUSSION THREAD

**Discussion by davidmurdoch** [ethereumjs/ethereumjs-monorepo#2569] [General]
Looking into Shanghai for Ganache and in reading [EIP 3860](https://eips.ethereum.org/EIPS/eip-3860) it isn't clear to me if the raw `transaction.data` field should be subject to the limits of initcode, as transaction data isn't the same thing as `initcode`. Ethereumjs tx package _does_ check the length of the raw transaction data: https://github.com/ethereumjs/ethereumjs-monorepo/blob/1fbc0772a277a8a892d792f568e65f9e1ca9c4bd/packages/tx/src/eip2930Transaction.ts#L173 .

**Comment by davidmurdoch** [comment]
We've talked to the hyperledger besu team and they believe the new EIP 3860 `initcode` limit should only be considered during contract creation, and not used to limit the raw transaction data field limit.

https://discord.com/channels/905194001349627914/905205502940696607/1082835346829619241

Invite link: https://discord.gg/hyperledger

cc @jochem-brouwer @holgerd77

**Comment by jochem-brouwer** [comment]
What discord server is this? I cannot read this discussion or see what server it is.

It seems to me that you have indeed found a bug. To clarify: initcode max length check should only occur if there is a create transaction (`to` field is `null` / `undefined` / "empty"). However, if you send a "normal" transaction (i.e. to a contract with a big load of data) then this is not subject to this initcode size check.

**Reply by jochem-brouwer** [reply]
Note: this is my interpretation but I am rather sure that this is how the EIP is meant - it only covers contract creation context, not "normal" transactions to a contract.

**Reply by davidmurdoch** [reply]
Oh man I hate discord so much. Haha. It is a link to the hyperledger besu discord. I think this will work as an invite? https://discord.gg/hyperledger

Even in the case where the to field is empty, I think the `data` field is not all `initcode`. If an `initcode` size check was to happen in the Transaction constructors, like it is in the linked code above, I think the data would need to be parsed into its constituent parts of "init code" and "runtime code".

**Reply by jochem-brouwer** [reply]
I joined this discord (not sure why I was not in it yet), but can still not see this discussion. What is your discord handle? You can send me a message also, `Jochem#8569` is my handle.

`data` field when `to` field in a transaction is empty is actually initcode. This is a big gotcha in EVM: if your `to` field is empty, and you provide data in the field, then if you `CALLDATACOPY` (or anything related to `CALLDATA`) will always be empty. However, if you now `CODECOPY` (or anything `CODE`-related) then this actually reads from the data field.

**Reply by jochem-brouwer** [reply]
Nvm, can actually see the question now.

**Reply by jochem-brouwer** [reply]
I will join the discussion because I think this is not correct at Besu's side.

**Reply by jochem-brouwer** [reply]
See rule #1 of the EIP @davidmurdoch :

`If length of transaction data (initcode) in a create transaction exceeds MAX_INITCODE_SIZE, transaction is invalid. (Note that this is similar to transactions considered invalid for not meeting the intrinsic gas cost requirement.)`

**Reply by davidmurdoch** [reply]
Yeah, but initcode is not the same thing as transaction data (there is init code, runtime code, and params in the data field). I'm not suggesting one way is right and the other is wrong, just that the rule is contradictory.

I think the spirit of the EIP is to limit initcode size, so limiting transaction data doesn't seem to make sense.

**Reply by jochem-brouwer** [reply]
I think we actually both have the same idea, but just to clarify, the rules are (I think):

- If `to` field is `null`, data field has a limit (`MAX_INITCODE_SIZE`)
- If `to` field is not `null`, data field has no max size

Note: data field technically has a max size since it is limited by blocks' gas limit, since we charge for data size of the tx.

**Reply by davidmurdoch** [reply]
I'm suggesting that the data field in the case where to is empty is NOT just init code.

**Reply by jochem-brouwer** [reply]
I still think there is a misunderstanding here. `data` field is either initcode (create TX) or data field (default tx). There are no extra fields, it is not possible to have initcode + a data field in a create tx, it is also not possible to have initcode + data field in a "default" tx. The to field is essentially the switch: if `to` is empty, the data field is initcode, otherwise it is the data field. It is not possible to have both.

**Reply by jochem-brouwer** [reply]
Regarding this (very-most-likely bug), see: https://github.com/ethereumjs/ethereumjs-monorepo/pull/2575

**Reply by jochem-brouwer** [reply]
Feel free to contact me on Discord also :)

**Reply by davidmurdoch** [reply]
Do you know of a canonical source for an EIP-relative definition of `initcode`? https://www.rareskills.io/post/ethereum-contract-creation-code describes the `data` field for a contract creating transaction from, Solidity's perspective, as `<init code> <runtime code> <constructor parameters>`. And another blog, https://monokh.com/posts/ethereum-contract-creation-bytecode, describes it similarly.

On Stack Exchange https://ethereum.stackexchange.com/questions/76334/what-is-the-difference-between-bytecode-init-code-deployed-bytecode-creation describes `initcode === contract creation code === data field`.

I'm not taking a stance either way, I'm just trying to understand the terms.

**Reply by davidmurdoch** [reply]
I found an implementation of the initcode size check in geth, so if we want to treat their implementation as canonical, it would seem `initcode === tx.data`, with the empty `to` address check just as in your latest PR: https://github.com/ethereum/go-ethereum/blob/793f0f9ec860f6f51e0cec943a268c10863097c7/cmd/evm/internal/t8ntool/transaction.go#L174-L177

**Reply by jochem-brouwer** [reply]
Yes, this sounds/looks absolutely right to me.

Again, the gotcha here is: the fields `initdata` or `data` (or whatever you want to call it) of tx are the same fields. The `to` field determines the behavior:

- It is empty: create a contract
- It is not-empty: normal transaction

Please note: if you send a tx to the zero-address this is still a normal transaction, since `to` field is the zero address and not "non-zero". 

Hence, by 3860, check iff (if-and-only-if) `to` field is empty (zero-length buffer/null/undefined), if data field does not exceed `MAX_INITCODE_SIZE`. If it does, tx is invalid and it is rejected over all nodes/pools, and will never be included in any block. (This is covered by https://github.com/ethereumjs/ethereumjs-monorepo/pull/2575 )

**Comment by jochem-brouwer** [comment]
Side note: this should be covered by tests, and this should also be tested on a testnet.
